src/devices/cpu/i386/i386.cpp

```diff
@@ -3870,6 +3870,74 @@ void i386_device::device_reset()
 	CHANGE_PC(m_eip);
 }
 
+inline void i386_device::dri_changed(uint8_t dr)
+{
+	int dr_enabled = (m_dr[7] & (1 << (dr << 1))) || (m_dr[7] & (1 << ((dr << 1) + 1))); // Check both local enable AND global enable bits for this breakpoint.
+	if(dr_enabled)
+	{
+		if(m_dr_breakpoints[dr]) m_dr_breakpoints[dr]->remove();
+		int breakpoint_type = (m_dr[7] >> ((dr << 2) + 16)) & 3;
+		int breakpoint_length = (m_dr[7] >> ((dr << 2) + 16 + 2)) & 3;
+		uint32_t phys_addr = m_dr[dr];
+		uint32_t error;
+		phys_addr = translate_address(m_CPL, TRANSLATE_READ, &phys_addr, &error);
+		phys_addr &= ~3; // According to CUP386, data breakpoints are only reliable on dword-aligned addresses, so align this to a dword.
+		uint32_t true_mask = 0;
+		switch(breakpoint_length)
+		{
+			case 0: true_mask = 0xff; break;
+			case 1: true_mask = 0xffff; break;
+			// Case 2 is invalid on a real 386.
+			case 3: true_mask = 0xffffffff; break;
+		}
+		if(true_mask == 0)
+		{
+			logerror("i386: Unknown breakpoint length value\n");
+		}
+		else if(breakpoint_type == 1) m_dr_breakpoints[dr] = m_program->install_write_tap(phys_addr, phys_addr + 3, "i386_debug_write_breakpoint",
+		[&, dr, true_mask](offs_t offset, u32& data, u32 mem_mask)
+		{
+			if(true_mask & make_bitmask<uint32_t>(m_program->data_width()))
+			{
+				m_dr[6] |= 1 << dr;
+				i386_trap(1,0,0);
+			}
+		}, m_dr_breakpoints[dr]);
+		else if(breakpoint_type == 3) m_dr_breakpoints[dr] = m_program->install_readwrite_tap(phys_addr, phys_addr + 3, "i386_debug_readwrite_breakpoint",
+		[&, dr, true_mask](offs_t offset, u32& data, u32 mem_mask)
+		{
+			if(true_mask & make_bitmask<uint32_t>(m_program->data_width()))
+			{
+				m_dr[6] |= 1 << dr;
+				i386_trap(1,0,0);
+			}
+		},
+		[&, dr, true_mask](offs_t offset, u32& data, u32 mem_mask)
+		{
+			if(true_mask & make_bitmask<uint32_t>(m_program->data_width()))
+			{
+				m_dr[6] |= 1 << dr;
+				i386_trap(1,0,0);
+			}
+		}, m_dr_breakpoints[dr]);
+	}
+}
+
+inline void i386_device::dr7_changed(uint32_t old_val, uint32_t new_val)
+{
+	int breakpoint_types = (old_val >> 16) & 0x3333;
+	int breakpoint_lengths = (old_val >> 18) & 0x3333;
+	int new_breakpoint_types = (new_val >> 16) & 0x3333;
+	int new_breakpoint_lengths = (new_val >> 18) & 0x3333;
+	if(breakpoint_types != new_breakpoint_types || breakpoint_lengths != new_breakpoint_lengths)
```

Comment:
This is ((old_val & 0xffff0000) != (new_val & 0xffff0000)). You can add a comment that explains what it means, instead of taking it through verbose code.

Reply:
ah good point, thanks!